plugins/outputs/groundwork/groundwork.go

```diff
@@ -0,0 +1,363 @@
+	"github.com/influxdata/telegraf/plugins/outputs"
+)
+
+const (
+	defaultMonitoringRoute = "/api/monitoring?dynamic=true"
+)
+
+// Login and logout routes from Groundwork API
+const (
+	loginURL  = "/api/auth/login"
+	logoutURL = "/api/auth/logout"
+)
+
+var (
+	sampleConfig = `
+  ## URL of your groundwork instance.
+  url = "https://groundwork.example.com"
+
+  ## Agent uuid for GroundWork API Server.
+  agent_id = ""
+
+  ## Username and password to access GroundWork API.
+  username = ""
+  password = ""
+
+  ## Default display name for the host with services(metrics).
+  # default_host = "telegraf"
+
+  ## Default service state.
+  # default_service_state = "SERVICE_OK"
+
+  ## The name of the tag that contains the hostname.
+  # resource_tag = "host"
+`
+)
+
+type Groundwork struct {
+	Server              string `toml:"url"`
+	AgentID             string `toml:"agent_id"`
+	Username            string `toml:"username"`
+	Password            string `toml:"password"`
+	DefaultHost         string `toml:"default_host"`
+	DefaultServiceState string `toml:"default_service_state"`
+	ResourceTag         string `toml:"resource_tag"`
+	authToken           string
+
+	Log telegraf.Logger `toml:"-"`
+}
+
+func (g *Groundwork) SampleConfig() string {
+	return sampleConfig
+}
+
+func (g *Groundwork) Init() error {
+	if g.Server == "" {
+		return errors.New("no 'url' provided")
+	}
+	if g.AgentID == "" {
+		return errors.New("no 'agent_id' provided")
+	}
+	if g.Username == "" {
+		return errors.New("no 'username' provided")
+	}
+	if g.Password == "" {
+		return errors.New("no 'password' provided")
+	}
+	if g.DefaultHost == "" {
+		return errors.New("no 'default_host' provided")
+	}
+	if g.ResourceTag == "" {
+		return errors.New("no 'resource_tag' provided")
+	}
+	if !validStatus(g.DefaultServiceState) {
+		return errors.New("invalid 'default_service_state' provided")
+	}
+
+	return nil
+}
+
+func (g *Groundwork) Connect() error {
+	byteToken, err := login(g.Server+loginURL, g.Username, g.Password)
+	if err != nil {
+		return fmt.Errorf("could not log in at %s: %v", g.Server+loginURL, err)
+	}
+
+	g.authToken = string(byteToken)
+
+	return nil
+}
+
+func (g *Groundwork) Close() error {
+	formValues := map[string]string{
+		"gwos-app-name":  "telegraf",
+		"gwos-api-token": g.authToken,
+	}
+
+	headers := map[string]string{
+		"Content-Type": "application/x-www-form-urlencoded",
+		"User-Agent":   internal.ProductToken(),
+	}
+
+	if _, _, err := clients.SendRequest(http.MethodPost, g.Server+logoutURL, headers, formValues, nil); err != nil {
+		return fmt.Errorf("could not log out at %s: %v", g.Server+logoutURL, err)
+	}
+
+	return nil
+}
+
+func (g *Groundwork) Write(metrics []telegraf.Metric) error {
+	resourceToServicesMap := make(map[string][]transit.DynamicMonitoredService)
+	for _, metric := range metrics {
+		resource, service := g.parseMetric(metric)
+		resourceToServicesMap[resource] = append(resourceToServicesMap[resource], service)
```

Comment:
Can we be _really_ **sure** that `resource` exists in `resourceToServicesMap`? As otherwise we assign to `nil` which likely will panic...

Reply:
Will not panic. You can check https://go.dev/play/p/4vODJRRuEUo for proves